rfd/0091-non-ad-desktop-access.md

```diff
@@ -0,0 +1,144 @@
+
+Credential provider is registered by adding it to registry under keys:
+
+- `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Authentication\Credential Providers`
+- `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Authentication\Credential Provider Filters`
+
+### Authentication package
+
+Authentication package will implement minimal set on method required by Windows:
+
+- `LsaApInitializePackage` used to initialize package and provides table with necessary helper functions
+- `LsaApLogonUser` used for actual logon process - certificate verification and creation of token happens here
+- `LsaApCallPackage`, `LsaApCallPackagePassthrough`, `LsaApCallPackageUntrusted`, `LsaApLogonTerminated` - empty
+  implementations used in other scenarios, required in all authentication packages
+
+`LsaApLogonUser` will verify certificate using Go's [x509 package](https://pkg.go.dev/crypto/x509) against system Root
+CAs - Telport Root Certificate will have to be imported to system store. It will also verify challenge (
+see [Certificate verification](#certificate-verification)).
+
+If everything is correct it will ask SAM database for user information
+(
+using [LsaGetAuthDataForUser](https://learn.microsoft.com/en-us/windows/win32/api/ntsecpkg/nc-ntsecpkg-lsa_get_auth_data_for_user)
+,
+[LsaConvertAuthDataToToken](https://learn.microsoft.com/en-us/windows/win32/api/ntsecpkg/nc-ntsecpkg-lsa_convert_auth_data_to_token)
+,
+and [GetTokenInformation](https://learn.microsoft.com/en-us/windows/win32/api/securitybaseapi/nf-securitybaseapi-gettokeninformation))
+and will create logon token - similar thing is done by
+OpenSSH [LSA implementation](https://github.com/PowerShell/Win32-OpenSSH/blob/master/contrib/win32/win32compat/lsa/Ssh-lsa.cpp#L657)
+. It will also create profile buffer in a way compatible with
+default [MSV1_0 authentication package](https://learn.microsoft.com/en-us/windows/win32/secauthn/msv1-0-authentication-package)
+. Example of C implementation can be
+found [here](https://github.com/Paolo-Maffei/OpenNT/blob/5c5b979ec08c17d3ca2eb70e8aad62d26515d01c/ds/lsa/msv1_0/nlp.c#L622)
+.
+
+Authentication package is registered by adding it to registry under key:
+
+- `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Lsa\Authentication Packages`
+
+### Certificate verification
+
+To verify smart card certificate we need to
+use [wincrypt API](https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/). Using it from authentication provider
+is impossible - `lsass.exe` runs in different context and doesn't have access
+to [Microsoft Base Smart Card provider](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/secsmart/microsoft-base-smart-card-cryptographic-service-provider)
+. We have to obtain information to verify that user has correct certificate in credential provider. To do that we will
+sign current time + secret shared between credential provider and authentication package using smart card.
```

Comment:
Could you describe the signing process in more detail? Are we using the `Microsoft Base Smart Card provider` for the signing purposes?